main.tf

```diff
@@ -32,30 +32,59 @@ data "aws_subnet_ids" "private" {
     values = ["${var.network_tag}_ecs_private_subnet"]
   }
 }
-
+#######
+# IAM
+#######
 resource "aws_iam_role" "ECSTaskExecutionRole" {
   name_prefix = var.name_prefix
 
-  # Terraform's "jsonencode" function converts a
-  # Terraform expression result to valid JSON syntax.
-  assume_role_policy = jsonencode({
-    Version = "2012-10-17"
-    Statement = [
-      {
-        Action = "sts:AssumeRole"
-        Effect = "Allow"
-        Principal = {
-          Service = "ecs-tasks.amazonaws.com"
-        }
-      },
-    ]
-  })
+  assume_role_policy = data.aws_iam_policy_document.allow_assume_ecs_task.json
 
   tags = {
     Environment = "aws-ia-fargate"
   }
 }
 
+resource "aws_iam_role" "task_role" {
+  name_prefix = var.name_prefix
+
+  assume_role_policy = data.aws_iam_policy_document.allow_assume_ecs_task.json
+  tags = {
+    Environment = "aws-ia-fargate"
+  }
+}
+
+data "aws_iam_policy_document" "allow_assume_ecs_task" {
+  statement {
+    actions = ["sts:AssumeRole"]
+    principals {
+      identifiers = ["ecs-tasks.amazonaws.com"]
+      type        = "Service"
+    }
+  }
+}
+
+data "aws_iam_policy_document" "allow-ecs-exec" {
+  version = "2012-10-17"
+  statement {
+    effect = "Allow"
+    actions = [
+      "ssmmessages:CreateControlChannel",
+      "ssmmessages:CreateDataChannel",
+      "ssmmessages:OpenControlChannel",
+      "ssmmessages:OpenDataChannel"
+    ]
+    resources = ["*"]
+  }
+}
```

Comment:
Two things:

1. The * in the resources portion of this policy concerns me. 

2. Doing a little bit of digging I found this blog post: https://aws.amazon.com/blogs/containers/new-using-amazon-ecs-exec-access-your-containers-fargate-ec2/

I see the exact policy statement in this blog. I would like to see the other controls of auditability/logging alongside this kind of exec privilege.

Reply:
@gargana
__IAM roles__ No problem. It will be added. 

__Comment 1__ The documentation states: `Amazon Session Manager Message Gateway Service does not support specifying a resource ARN in the Resource element of an IAM policy statement. To allow access to Amazon Session Manager Message Gateway Service, specify “Resource”: “*” in your policy. `  [Source](https://docs.aws.amazon.com/service-authorization/latest/reference/list_amazonsessionmanagermessagegatewayservice.html) The article also used `resource "*"` so it's seems like it's not possible to limit it. 

__Comment 2__ Just so that we're on the same page with what additional controls you would like to see. You'd like Cloudwatch Logs and KMS support to be added?